packages/core/contracts/financial-templates/common/FundingRateApplier.sol

```diff
@@ -94,15 +153,65 @@ abstract contract FundingRateApplier is Testable, Lockable {
         view
         returns (FixedPoint.Unsigned memory tokenDebt)
     {
-        return rawTokenDebt.mul(cumulativeFundingRateMultiplier);
+        return rawTokenDebt.mul(fundingRate.cumulativeMultiplier);
     }
 
-    function _getFundingRateStore() internal view returns (FundingRateStoreInterface) {
-        return FundingRateStoreInterface(finder.getImplementationAddress("FundingRateStore"));
+    function _getOptimisticOracle() internal view returns (OptimisticOracle) {
+        return OptimisticOracle(finder.getImplementationAddress(OracleInterfaces.OptimisticOracle));
     }
 
-    function _getLatestFundingRate() internal view returns (FixedPoint.Signed memory) {
-        return _getFundingRateStore().getFundingRateForContract(address(this));
+    // function priceSettled(
+    //     bytes32 identifier,
+    //     uint256 timestamp,
+    //     int256 price
+    // ) external nonReentrant() {
+    //     // Check to be sure this is a valid request that is NOT the currently proposed one.
+    //     require(msg.sender == address(_getOptimisticOracle()) && identifier == fundingRate.identifier && timestamp != fundingRate.proposalTime && timestamp > fundingRate.updateTime, "Irrelevant settlement");
+
+    //     // Set the update time and the rate. This doesn't affect any currently proposed request.
+    //     fundingRate.updateTime = timestamp;
+    //     fundingRate.rate = FixedPoint.Signed(price);
+    // }
+
+    function _getLatestFundingRate() internal returns (FixedPoint.Signed memory) {
+        uint256 timestamp = fundingRate.proposalTime;
+        if (timestamp != 0) {
+            // Attempt to update the funding rate.
+            OptimisticOracle optimisticOracle = _getOptimisticOracle();
+            bytes32 identifier = fundingRate.identifier;
+
+            // Try to get the price from the optimistic oracle.
+            try optimisticOracle.getPrice(identifier, timestamp) returns (int256 price) {
+                // If successful, figure out the type of request.
+                OptimisticOracle.Request memory request =
+                    optimisticOracle.getRequest(address(this), identifier, timestamp);
+                uint256 lastUpdateTime = fundingRate.updateTime;
+
+                // If the request is more recent than the last update then we should update the funding rate.
+                if (timestamp >= lastUpdateTime) {
+                    // Update funding rates
+                    fundingRate.rate = FixedPoint.Signed(price);
+                    fundingRate.updateTime = timestamp;
+
+                    // If there was no dispute, send a reward.
+                    if (request.disputer == address(0)) {
+                        FixedPoint.Unsigned memory reward =
+                            _pfc().mul(fundingRate.rewardRate).mul(timestamp.sub(lastUpdateTime));
```

Comment:
Thinking through weird scenarios, is there ever a situation where `timestamp.sub(lastUpdateTime) < 0`?

Reply:
I had the exact same question -- that's why I put this inside the outer `if` that checks that condition, so this whole update would be skipped in that case.